src/ca65/scanner.c

```diff
@@ -403,29 +404,111 @@ static void IFNextChar (CharSource* S)
-        ** with strange line terminators, remove all whitespace from the
-        ** end of the line, the add a single newline.
+        ** with strange line terminators, count and remember the position of 
+        ** trailing whitespace characters, but keep the line intact to
+        ** support raw input.
         */
-        Len = SB_GetLen (&S->V.File.Line);
-        while (Len > 0 && IsSpace (SB_AtUnchecked (&S->V.File.Line, Len-1))) {
+
+        while (Len > 0 && IsSpace (SB_AtUnchecked (&S->V.File.Line, Len - 1))) {
             --Len;
         }
-        SB_Drop (&S->V.File.Line, SB_GetLen (&S->V.File.Line) - Len);
-        SB_AppendChar (&S->V.File.Line, '\n');
-
-        /* Terminate the string buffer */
-        SB_Terminate (&S->V.File.Line);
+        S->V.File.EndLinePos = Len;
 
         /* One more line */
         S->V.File.Pos.Line++;
 
         /* Remember the new line for the listing */
         NewListingLine (&S->V.File.Line, S->V.File.Pos.Name, FCount);
+    }
+
+    /* Set the column pointer */
+    S->V.File.Pos.Col = SB_GetIndex (&S->V.File.Line);
+
+    /* Return the next character:
+    ** Check for special case: Are we at the begining of whitespace?
+    ** If so, return EOL.
+    */
+    if (SB_GetIndex (&S->V.File.Line) == S->V.File.EndLinePos) {
+        C = '\n';
+        /* Increment the index */
+        SB_SetIndex (&S->V.File.Line, SB_GetIndex (&S->V.File.Line) + 1);
+    } else {
+        /* Return the next character from the buffer */
+        C = SB_Get (&S->V.File.Line);
+    }
+
+}
+
+
+
+static void IFNextRawChar (CharSource* S)
+/* Read the next raw character from the input file */
+{
+    
+    /* Set the column pointer */
+    S->V.File.Pos.Col = SB_GetIndex (&S->V.File.Line);
+        
+    /* Return the next character from the buffer, unless we are at the actual end. */
+    C = SB_Get (&S->V.File.Line);
+
+    if (C != 0) return;
+
+    /* If we come here, end of current line reached, read next line: */
+
+    SB_Clear (&S->V.File.Line);
+    while (1) {
+
+        int N = fgetc (S->V.File.F);
+        if (N == EOF) {
+            /* End of file. Accept files without a newline at the end */
+            if (SB_NotEmpty (&S->V.File.Line)) {
+                break;
+            }
+
+            /* No more data - add an empty line to the listing. This
+            ** is a small hack needed to keep the PC output in sync.
+            */
+            NewListingLine (&EmptyStrBuf, S->V.File.Pos.Name, FCount);
+            C = EOF;
+            return;
+
+        /* Check for end of line */
+        } else if (N == '\n') {
+
+            /* End of line */
+            break;
 
+        /* Collect other stuff */
+        } else {
+
+            /* Append data to line */
+            SB_AppendChar (&S->V.File.Line, N);
+
+        }
     }
 
+    /* We didn't copy the EOL in the while loop, so add it now */
+    SB_AppendChar (&S->V.File.Line, '\n');
```

Comment:
You should copy the EOL in the while loop.  That is, you should "collect stuff" before you test for EOL.